### PR TITLE
Implements a ProbabilityGridPointsProcessor

### PR DESCRIPTION
**Pull request by damienrg**
Hi,
this is wip on issue #355 and for now it should compile but it does not work and I need to do some cleanup.
For now, could you just tell me if:

- the name of the class are ok
- should I use a probability grid inside ImagePointsProcessor? If yes I do not know how to use it correctly so could you elaborate on this?
- in ProbabilityGridPointsProcessor, how should I create the ProbabilityGrid?

### CODE REVIEW

**Comment by damienrg** [comment]
I signed it!

**Comment by SirVer** [comment]
> the name of the class are ok

ImagePointsProcessor seems unnecessary. I think you should concentrate on ProbabilityGridPointsProcessor.

> should I use a probability grid inside ImagePointsProcessor? If yes I do not know how to use it correctly so could you elaborate on this?

Yes, I think you want to use it. Have a look at #387. We just implemented that the ProbabilityGrid can be resized. So you would create it at an arbitrary size, have something like `GrowAsNeeded` in your points processor when you process points. Inside `Flush` you would [ComputeCroppedProbabilityGrid](https://github.com/googlecartographer/cartographer/blob/a313448bf01cc10f015ad872098d58b5b9e5162c/cartographer/mapping_2d/submaps.cc#L33) so the white space is removed. This is the data you want to serialize. Now there are many useful ways of serializing it:

1) As a PNG. I suggest doing this first. Please share code with XRayPointsProcessor as much as possible. Also keep in mind that we'd like to draw the trajectory onto this PNG eventually, so you need to keep some understanding how your cropped pixels relate to the map frame.
2) As PGM/YAML for ROS. Once 1) is done, a followup PR can add a PointsProcessor to cartographer_ros, that writes out a YAML and a PGM for the ROS map server. This code would reuse the implementation of [WriteOccupancyGridToPgmAndYaml](https://github.com/googlecartographer/cartographer_ros/blob/764004736e49db70ab8a8d0d73189deedab98c7c/cartographer_ros/cartographer_ros/map_writer.cc#L82) to get to ROS like data.
3) As a proto. I am not sure if anybody cares about the proto format for now, so let's skip this.

> in ProbabilityGridPointsProcessor, how should I create the ProbabilityGrid?

I hope that is answered through #387.

**Comment by damienrg** [comment]
For now I do not handle the trajectory as:
1. I am not sure that the current implementation fit with what you have in mind
2. I need to find how to do it. For now I think that I should start by adding proto::Trajectory like it is done in XRayPointsProcessor.

File: cartographer/io/image_utils.cc [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#include "cartographer/io/image_utils.h"
```
Comment (SirVer) [inline cartographer/io/image_utils.cc]: I do not agree with *utils.h name of this file. How about just calling it Image?

File: cartographer/io/image_utils.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+#include <memory>
+
+#include "cairo/cairo.h"
+
```
Comment (SirVer) [inline cartographer/io/image_utils.cc]: remove empty line between includes

File: cartographer/io/image_utils.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+int stride_for_width(int width)
+{
+  const int stride = cairo_format_stride_for_width(kCairoFormat, width);
+  CHECK_EQ(stride % 4, 0);
+  return stride;
+}
+
+}  // namespace
+
+Image CreateImage(int width, int height)
```
Comment (SirVer) [inline cartographer/io/image_utils.cc]: It feels like this should now be a class `Image` with a constructor. You can add pixel accessors and hide all of the Cairo implementation in this file.

File: cartographer/io/image_utils.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#ifndef CARTOGRAPHER_IO_IMAGE_UTILS_H_
+#define CARTOGRAPHER_IO_IMAGE_UTILS_H_
+
+#include <cstdint>
+#include <vector>
+
+namespace cartographer {
+namespace io {
+
+class FileWriter;
```
Comment (SirVer) [inline cartographer/io/image_utils.h]: do not forward declare: https://google.github.io/styleguide/cppguide.html#Forward_Declarations

File: cartographer/io/image_utils.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{
+  int width;
+  int height;
+  int stride;
+  std::vector<uint32_t> pixels;
+};
+
+Image CreateImage(int width, int height);
+
+void WritePng(Image * const image, FileWriter * const file_writer);
```
Comment (SirVer) [inline cartographer/io/image_utils.h]: this can move into the Image class then.

File: cartographer/io/probability_grid_points_processor.cc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "cartographer/common/make_unique.h"
+#include "cartographer/common/math.h"
+#include "cartographer/io/image_utils.h"
+
+namespace cartographer {
+namespace io {
+namespace {
+
+struct PixelData {
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: Use `Color` typedef from points_batch.h instead?

File: cartographer/io/probability_grid_points_processor.cc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  uint8_t r = 0;
+  uint8_t g = 0;
+  uint8_t b = 0;
+};
+
+using PixelDataMatrix =
+  Eigen::Matrix<PixelData, Eigen::Dynamic, Eigen::Dynamic>;
+
+cairo_status_t CairoWriteCallback(void* const closure,
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: unused? Remove!

File: cartographer/io/probability_grid_points_processor.cc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      image.pixels[y * image.stride / 4 + x] = (255 << 24) | (r << 16) | (g << 8) | b;
+    }
+  }
+
+  WritePng(&image, file_writer);
+  CHECK(file_writer->Close());
+}
+
+mapping_2d::ProbabilityGrid CreateProbabilityGrid(double resolution) {
+  constexpr int kInitialProbabilityGridSize = 100;
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: What a terrible interface ProbabilityGrid provides for construction. I am sorry .... :(.

File: cartographer/io/probability_grid_points_processor.cc [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+  common::LuaParameterDictionary* const dictionary,
+  PointsProcessor* const next) {
+  return common::make_unique<ProbabilityGridPointsProcessor>(
+    dictionary->GetDouble("resolution"),
+      mapping_2d::CreateRangeDataInserterOptions(
+          dictionary->GetDictionary("range_data_inserter").get()),
+    file_writer_factory(dictionary->GetString("filename") + ".png"), next);
+}
+
+void WriteGrid(const mapping_2d::ProbabilityGrid & probability_grid,
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: Can be moved into anonymous namespace?

File: cartographer/io/probability_grid_points_processor.cc [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    return Eigen::Array2i(cell_limits.num_x_cells - index(0) - 1,
+                          cell_limits.num_y_cells - index(1) - 1);
+  };
+  int width = cell_limits.num_y_cells;
+  int height = cell_limits.num_x_cells;
+  PixelDataMatrix image(height, width);
+  for (auto xy_index :
+         cartographer::mapping_2d::XYIndexRangeIterator(cell_limits)) {
+    auto index = xy_index + offset;
+    if (probability_grid.IsKnown(index)) {
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: Please write constexpr uint8_t kUnknownValue = 128; for unknown space to stay consistent with `cartographer_ros/map_writer.cc`.

File: cartographer/io/probability_grid_points_processor.cc [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      const Eigen::Array2i pixel = grid_index_to_pixel(xy_index);
+      PixelData& pixel_data = image(pixel.x(), pixel.y());
+      float probability = probability_grid.GetProbability(index);
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: can be const. Also to stay consistent, this should be `1.f - probability_grid...`.

File: cartographer/io/probability_grid_points_processor.cc [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+      uint8_t value = 255
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: `p - min / (max - min)` should read `(p - min) / (max - min)`

File: cartographer/io/probability_grid_points_processor.cc [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        * (probability - mapping::kMinProbability
+           / (mapping::kMaxProbability - mapping::kMinProbability));
+      pixel_data.r = value;
+      pixel_data.g = value;
+      pixel_data.b = value;
+    }
+  }
+
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.cc]: remove empty line

File: cartographer/io/probability_grid_points_processor.h [hunk 14]
```diff
@@ -0,0 +1,56 @@
+#ifndef CARTOGRAPHER_IO_PROBABILITY_GRID_POINTS_PROCESSOR_H_
+#define CARTOGRAPHER_IO_PROBABILITY_GRID_POINTS_PROCESSOR_H_
+
+// Library used for inserting range data points into a probability grid.
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.h]: comment doesn't add a ton IMHO. remove?

File: cartographer/io/probability_grid_points_processor.h [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+#include "cartographer/mapping_2d/probability_grid.h"
+#include "cartographer/mapping_2d/proto/range_data_inserter_options.pb.h"
+#include "cartographer/mapping_2d/range_data_inserter.h"
+
+namespace cartographer {
+namespace io {
+
+// Creates a probability grid with the specified 'resolution'.
+// 'range_data_inserter' options are used to configure the range
+// data ray tracing through the probability grid.
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.h]: Mention that the `z` component of the data is ignored, i.e. all points are projected into the x-y plane.

File: cartographer/io/probability_grid_points_processor.h [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+  ~ProbabilityGridPointsProcessor() override {}
+
+  void Process(std::unique_ptr<PointsBatch> batch) override;
+  FlushResult Flush() override;
+
+ private:
+  PointsProcessor* const next_;
+  std::unique_ptr<FileWriter> file_writer_;
+
```
Comment (SirVer) [inline cartographer/io/probability_grid_points_processor.h]: remove empty line?

File: cartographer/io/image_utils.cc [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  if (static_cast<FileWriter*>(closure)->Write(
+        reinterpret_cast<const char*>(data), length)) {
+    return CAIRO_STATUS_SUCCESS;
+  }
+  return CAIRO_STATUS_WRITE_ERROR;
+}
+
+constexpr cairo_format_t kCairoFormat = CAIRO_FORMAT_ARGB32;
+
+int stride_for_width(int width)
```
Comment (SirVer) [inline cartographer/io/image_utils.cc]: GetStrideForWidth

File: cartographer/io/image_utils.h (hunk 4 above)
Comment (damienrg) [inline cartographer/io/image_utils.h]: Ok. I see that you seem to repeat includes from the header in the source files. Should I do the same?

**Comment by damienrg** [comment]
In `ProbabilityGridPointsProcessor::Flush() ` do you think that the message of  `LOG(FATAL)...` should contain probability grid instead of image?

File: cartographer/io/image_utils.h (hunk 4 above)
Comment (SirVer) [inline cartographer/io/image_utils.h]: We are not consistent about that. It is enough to include the file in the header.

**Comment by SirVer** [comment]
Damien, thanks for the contribution! I fixed some minor nits and ran clang-format and then merged this. If you disagree with any of my changes we can discuss and change them in a separate PR again.

I see two followups for this PR: 

1) The point 2. I mentioned [above](https://github.com/googlecartographer/cartographer/pull/383#issuecomment-313083104) so that the occupancy grids can be used with map_server for ROS nodes. This would be a PR for `cartographer_ros`. 
2) Drawing the trajectory into the image. I think I wrote something about this already, but GitHub seems to have eaten it? I was musing that you'd need to take the Trajectories similar to how XRayPointsProcessor does it and that it would be cool to write out an SVG or TIF or something else that supports layers, so that the bottom layer would just be the map, the layer above the Trajectories and so on.

Are you interested in tackling any of these?

**Comment by damienrg** [comment]
> If you disagree with any of my changes we can discuss and change them in a separate PR again.

No problem with your changes.

> 1. The point 2. I mentioned [above](https://github.com/googlecartographer/cartographer/pull/383#issuecomment-313083104) so that the occupancy grids can be used with map_server for ROS nodes. This would be a PR for cartographer_ros.

To be sure I understand what you have in mind, the purpose is to be able to use it in the pipeline description file, no?

> 2. Drawing the trajectory into the image. I think I wrote something about this already, but GitHub seems to have eaten it? I was musing that you'd need to take the Trajectories similar to how XRayPointsProcessor does it and that it would be cool to write out an SVG or TIF or something else that supports layers, so that the bottom layer would just be the map, the layer above the Trajectories and so on.

For writing a SVG or TIF, you mean by hand or using a library? I do not know TIF but for SVG it is doable (embed the map in png and add a trajectory layer).

> Are you interested in tackling any of these?

Why not and I am more interested in point 2 but as I have less free time to work on these I cannot guarantee that it will be done soon.

**Comment by SirVer** [comment]
> For writing a SVG or TIF, you mean by hand or using a library? 

A library. We are using Cairo already and it does support SVG (https://www.cairographics.org/manual/cairo-SVG-Surfaces.html), not sure if it does support TIF and also not sure if it does support individual layers in images. That will require some research.

**Comment by damienrg** [comment]
> A library. We are using Cairo already and it does support SVG (https://www.cairographics.org/manual/cairo-SVG-Surfaces.html), not sure if it does support TIF and also not sure if it does support individual layers in images. That will require some research.

It is only for rendering SVG, not writting.

**Comment by SirVer** [comment]
> It is only for rendering SVG, not writing.

No, it does: use [cairo_svg_surface_create_for_stream](https://www.cairographics.org/manual/cairo-SVG-Surfaces.html) and use it similar to cairo_surface_write_to_png_stream.

**Pull request merged**